Build rb-cool.io

oblute opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams.
Consider asking on our [Gitter channel](https://gitter.im/conda-forge/conda-forge.github.io)
if your recipe isn't reviewed promptly.
-->

Checklist

- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [X] Source is from official source
- [X] Package does not vend other packages
- [X] Build number is 0
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/rb-cool.io```) and found it was in an excellent condition.

oblute commented:
@benhuff @sodre agree to be maintainers?

benhuff commented:
yes 👍

oblute commented:
@conda-forge/help-ruby this recipe is ready for your review

sodre commented on recipes/rb-cool.io/meta.yaml:
```diff
@@ -0,0 +1,45 @@
+  host:
+    - ruby
+  run:
+    - ruby
```
It looks like both `cool.io` and `nio4r` have a dependency on `libev` and are venturing the libraries. 

Lets wait until #9450 is fixed to understand how to properly unvendor libev from nio4r and apply the same technique to your package.

sodre replied:
@conda-forge/staged-recipes, @conda-forge/core

What is the guidance when vendoring patched versions of C libraries? 
In the case of `rb-cool.io`, it vendors a [patched](https://github.com/tarcieri/cool.io/blob/0fd3fd1d8e8d81e24f79f809979367abc3f52b92/libev_ruby_gil.diff#L113-L145) version of `libev`  to solve an incompatibility between Ruby's global-interpreter-lock and libev's concurrency management.

isuruf replied:
If it's absolutely necessary, then the vendored library and the original library must be co-installable and vendored library's license file must be added as a license in this package.

sodre replied:
Fair, any guidelines on how I can ensure both libraries can be co-installed?

isuruf replied:
Does this package install a `libev.so`?

sodre replied:
It does not look like it does, instead they include `libev.c` in their own extension. see 

https://github.com/tarcieri/cool.io/blob/master/ext/cool.io/libev.c

I am building the package locally to verify. I imagine, that if no libev.so is created that they can both be installed without conflicts, is that correct?

isuruf replied:
yes

oblute commented:
@sodre i pushed a potential fix for the vendoring of libev - i am entirely unsure if it actually remedied it, but it looks to be...let me know.

sodre commented:
@oblute, your changes are in the correct direction and necessary, but not sufficient to un-vendor libev. Here is what I think is sufficient:
  - [x] add libev as host and run requirements.
  - [ ] Unpack the .gem
  - [ ] Apply a patch to remove libev from the .gem, and ensure what is left points to libev under  ${PREFIX}/include and ${PREFIX}/lib
    Note: I believe you created the patch. However, it is hard for me to say what it does because it is very long. I suggest creating a patch that just fixes the cool.io's code and use `rm .../ext/libev` to remove the libev code.
  - [ ] Repack the .gem 
  - [ ] Gem install the repacked gem

These steps are more than what we should have within the `script` section of meta.yaml. So, lets follow @scopatz lead and move the code into build.sh.

oblute commented:
@conda-forge/help-ruby @scopatz is there any guidance I can get on moving the build script into a build.sh file for the above? i'm not familiar with writing build.sh files and unsure about the changes necessary to do so

oblute commented:
@sodre i believe the above implemented your list of fixes.

[x] added libev as a host dependency - not necessary for run.
[x] in a build.sh script, unpacked the gem, removed libev folder, cleaned .o files and repacked/installed gem
[x] created a patch that fixed the two spots in the cool.io code referencing the libev deleted folder and PREFIX points to what is left in the build.sh

oblute commented:
@sodre @conda-forge/help-ruby any thoughts on the proposed changes?
